File: km/hungarian.py
class Hungarian:
    def __init__(self, points, weights):
        self.points = points
        self.weights = weights

    #https://blog.csdn.net/c20180630/article/details/70175814
    #split nodes into X,Y List
    #points:list
    #weights:map<point, map<point, weight> >
    def bigraph(self):
        if len(points) <= 0: return [], []
        X = []
        Y = []
        #numbering
        global numbers
        numbers = {}
        for i in range(0, len(points)):
            point = points[i]
            if point in numbers: continue
            numbers[point] = 1
            if point not in weights: continue
            #dfs numbering
            def bigraph_dfs(point, weights, num):
                for p,w in weights[point].items():
                    if p in numbers: continue
                    if w <= 0: continue
                    numbers[p] = num + 1
                    bigraph_dfs(p, weights, num + 1)
            bigraph_dfs(point, weights, 1)
        #split
        for p in points:
            if numbers[p] % 2 == 0:
                Y.append(p)
            else:
                X.append(p)
        return X,Y

    #hungarian algorithm with dfs search
    #https://blog.csdn.net/c20180630/article/details/70175814
    #input:1.X, 2.weights
    #output:1.match_num, 2.matched_map
    def hungarian_dfs(self):
        X,Y = self.bigraph()
        global marked_Y
        global matched_Y
        marked_Y = []
        matched_Y = {}
        matched_num = 0
        for x in X:
            marked_Y = []
            #find augmented path
            def hungarian_dfs_process(x, weights):
                if x not in weights:
                    return 0
                res = 0
                for y,w in weights[x].items():
                    if y in marked_Y: continue
                    if y not in matched_Y:
                        matched_Y[y] = x
                        marked_Y.append(y)
                        return 1
                    else:
                        t = matched_Y[y]
                        matched_Y[y] = x
                        marked_Y.append(y)
                        res += hungarian_dfs_process(t, weights)
                        if res > 0:
                            break
                        else:
                            matched_Y[y] = t
                return res
            matched_num += hungarian_dfs_process(x, weights)
        return matched_num, matched_Y


points = ['x1', 'x2', 'x3', 'x4', 'y1', 'y2', 'y3', 'y4']
weights = {}
weights['x1'] = {'y1':1, 'y2':1}
weights['x2'] = {'y1':1, 'y3':1}
weights['x3'] = {'y2':1, 'y4':1}
weights['x4'] = {'y2':1, 'y3':1}
weights['y1'] = {'x1':1, 'x2':1}
weights['y2'] = {'x1':1, 'x3':1, 'x4':1}
weights['y3'] = {'x2':1, 'x4':1}
weights['y4'] = {'x3':1}
hungarian_process = Hungarian(points, weights)
print hungarian_process.hungarian_dfs()

File: km/hungarian.java
import java.util.*;

class Hungarian {
    private Map<String, Integer> m_numbers = new HashMap<String, Integer>();
    private List<String> m_marked_Y = new ArrayList<String>();
    private Map<String, String> m_matched_Y = new HashMap<String, String>();

    private void bigraph_dfs(String point, Map<String, Map<String, Float>> weights, int num){
        for(Map.Entry<String, Float> x : weights.get(point).entrySet()){
            if(m_numbers.containsKey(x.getKey())) continue;
            if(x.getValue() <= 0) continue;
            m_numbers.put(x.getKey(), num + 1);
            bigraph_dfs(x.getKey(), weights, num + 1);
        }
    }

    public void bigraph(List<String> points, Map<String, Map<String, Float>> weights, List<String> graph_X, List<String> graph_Y){
        if(points.size() <= 0) return ;
        m_numbers.clear();
        graph_X.clear();
        graph_Y.clear();
        //numbering
        for(int i=0; i<points.size(); i++){
            String point = points.get(i);
            if(m_numbers.containsKey(point)) continue;
            m_numbers.put(point, 1);
            if(! weights.containsKey(point)) continue;
            //dfs numbering
            bigraph_dfs(point, weights, 1);
        }
        //split
        for(Map.Entry<String, Integer> x: m_numbers.entrySet()){
            if(x.getValue() % 2 == 0){
                graph_X.add(x.getKey());
            }else{
                graph_Y.add(x.getKey());
            }
        }
    }

    private int hungarian_dfs_process(String x, Map<String, Map<String, Float>> weights){
        if(!weights.containsKey(x)) return 0;
        int res = 0;
        for(Map.Entry<String, Float> p : weights.get(x).entrySet()){
            String y = p.getKey();
            if(m_marked_Y.contains(y)) continue;
            if(!m_matched_Y.containsKey(y)){
                m_matched_Y.put(y, x);
                m_marked_Y.add(y);
                return 1;
            }else{
                String t = m_matched_Y.get(y);
                m_matched_Y.put(y, x);
                m_marked_Y.add(y);
                res += hungarian_dfs_process(t, weights);
                if(res > 0) break;
                else m_matched_Y.put(y, t);
            }
        }
        return res;
    }

    public Map<String, String> hungarian_dfs(List<String> points, Map<String, Map<String, Float>> weights){
        Map<String, String> matched_results = new HashMap<String, String>();
        List<String> X = new ArrayList<String>();
        List<String> Y = new ArrayList<String>();

        bigraph(points, weights, X, Y);

        for(String x : X){
            m_marked_Y.clear();
            hungarian_dfs_process(x, weights);
        }
        return m_matched_Y;
    }

    public static void main(String[] args){
        List<String> points = Arrays.asList("x1", "x2", "x3", "x4", "y1", "y2", "y3", "y4");
        Map<String, Map<String, Float>> weights = new HashMap<String, Map<String, Float>>();
        for(String s : points){
            weights.put(s, new HashMap<String, Float>());
        }
        weights.get("x1").put("y1", 1F);
        weights.get("x1").put("y2", 1F);
        weights.get("x2").put("y1", 1F);
        weights.get("x2").put("y3", 1F);
        weights.get("x3").put("y2", 1F);
        weights.get("x3").put("y4", 1F);
        weights.get("x4").put("y2", 1F);
        weights.get("x4").put("y3", 1F);
        weights.get("y1").put("x1", 1F);
        weights.get("y1").put("x2", 1F);
        weights.get("y2").put("x1", 1F);
        weights.get("y2").put("x3", 1F);
        weights.get("y2").put("x4", 1F);
        weights.get("y3").put("x2", 1F);
        weights.get("y3").put("x4", 1F);
        weights.get("y4").put("x3", 1F);
        Hungarian hungarian = new Hungarian();
        Map<String, String> res = hungarian.hungarian_dfs(points, weights);
        for (Map.Entry<String, String> p : res.entrySet()){
            System.out.println(p.getKey() + "--" + p.getValue());
        }
    }
}
